Index: build/obexd/src/manager.c
===================================================================
--- build.orig/obexd/src/manager.c
+++ build/obexd/src/manager.c
@@ -494,7 +494,7 @@ gboolean manager_init(void)
 
 	dbus_error_init(&err);
 
-	connection = g_dbus_setup_bus(DBUS_BUS_SESSION, OBEXD_SERVICE, &err);
+	connection = g_dbus_setup_bus(DBUS_BUS_SYSTEM, OBEXD_SERVICE, &err);
 	if (connection == NULL) {
 		if (dbus_error_is_set(&err) == TRUE) {
 			fprintf(stderr, "%s\n", err.message);
Index: build/obexd/client/ftp.c
===================================================================
--- build.orig/obexd/client/ftp.c
+++ build/obexd/client/ftp.c
@@ -476,7 +476,7 @@ int ftp_init(void)
 
 	DBG("");
 
-	conn = dbus_bus_get(DBUS_BUS_SESSION, NULL);
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
 	if (!conn)
 		return -EIO;
 
Index: build/obexd/client/map.c
===================================================================
--- build.orig/obexd/client/map.c
+++ build/obexd/client/map.c
@@ -2074,7 +2074,7 @@ int map_init(void)
 
 	DBG("");
 
-	conn = dbus_bus_get(DBUS_BUS_SESSION, NULL);
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
 	if (!conn)
 		return -EIO;
 
Index: build/obexd/client/opp.c
===================================================================
--- build.orig/obexd/client/opp.c
+++ build/obexd/client/opp.c
@@ -190,7 +190,7 @@ int opp_init(void)
 
 	DBG("");
 
-	conn = dbus_bus_get(DBUS_BUS_SESSION, NULL);
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
 	if (!conn)
 		return -EIO;
 
Index: build/obexd/client/pbap.c
===================================================================
--- build.orig/obexd/client/pbap.c
+++ build/obexd/client/pbap.c
@@ -1313,7 +1313,7 @@ int pbap_init(void)
 
 	DBG("");
 
-	conn = dbus_bus_get(DBUS_BUS_SESSION, NULL);
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
 	if (!conn)
 		return -EIO;
 
Index: build/obexd/client/session.c
===================================================================
--- build.orig/obexd/client/session.c
+++ build/obexd/client/session.c
@@ -593,7 +593,7 @@ struct obc_session *obc_session_create(c
 	if (driver == NULL)
 		return NULL;
 
-	conn = dbus_bus_get(DBUS_BUS_SESSION, NULL);
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
 	if (conn == NULL)
 		return NULL;
 
Index: build/obexd/client/sync.c
===================================================================
--- build.orig/obexd/client/sync.c
+++ build/obexd/client/sync.c
@@ -237,7 +237,7 @@ int sync_init(void)
 
 	DBG("");
 
-	conn = dbus_bus_get(DBUS_BUS_SESSION, NULL);
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
 	if (!conn)
 		return -EIO;
 
Index: build/tools/obexctl.c
===================================================================
--- build.orig/tools/obexctl.c
+++ build/tools/obexctl.c
@@ -2519,7 +2519,7 @@ int main(int argc, char *argv[])
 	}
 
 	main_loop = g_main_loop_new(NULL, FALSE);
-	dbus_conn = g_dbus_setup_bus(DBUS_BUS_SESSION, NULL, NULL);
+	dbus_conn = g_dbus_setup_bus(DBUS_BUS_SYSTEM, NULL, NULL);
 
 	rl_attempted_completion_function = cmd_completion;
 
